WooCommerce/Classes/Authentication/Store Creation/StoreCreationWebViewModel.swift

```diff
@@ -0,0 +1,94 @@
+import Foundation
+import WebKit
+
+/// View model used for the web view controller to create a store.
+///
+final class StoreCreationWebViewModel: AuthenticatedWebViewModel {
+    // `AuthenticatedWebViewModel` protocol conformance.
+    let title = Localization.title
+    let initialURL: URL? = Constants.storeCreationURL
+
+    private let completion: (Result<String, Error>) -> Void
+
+    init(completion: @escaping (Result<String, Error>) -> Void) {
+        self.completion = completion
+    }
+
+    func handleDismissal() {
+        // no-op: dismissal is handled in the close button in the navigation bar.
+    }
+
+    func handleRedirect(for url: URL?) {
+        guard let path = url?.absoluteString else {
+            return
+        }
+        handleCompletionIfPossible(path)
+    }
+
+    func decidePolicy(for navigationURL: URL) async -> WKNavigationActionPolicy {
+        handleCompletionIfPossible(navigationURL.absoluteString)
+        return .allow
+    }
+}
+
+enum StoreCreationError: Error {
+    case noSiteURLInCompletionPath
+    case invalidCompletionPath
+    case newSiteUnavailable
+    case newSiteIsNotJetpackSite
+}
+
+private extension StoreCreationWebViewModel {
+    func handleCompletionIfPossible(_ url: String) {
+        guard url.starts(with: Constants.completionURLPrefix) else {
+            return
+        }
+        do {
+            // A successful URL looks like `https://wordpress.com/checkout/thank-you/{{site_url}}/.*`.
+            // There is usually more than one URL requests like this, with different parameters.
+            let regex = try NSRegularExpression(pattern: "\(Constants.completionURLPrefix)" + #"(?<siteURL>[^/]+)"#, options: [])
+            let urlRange = NSRange(location: 0, length: url.count)
+            let matches = regex.matches(in: url, options: [], range: urlRange)
+            guard let match = matches.first else {
+                throw StoreCreationError.invalidCompletionPath
+            }
+
+            // Extracts the site URL substring matching the named capture group `siteURL` in the regex.
```

Comment:
Nit: Would it make sense to move the site URL extraction code to a private method?

Reply:
Updated in https://github.com/woocommerce/woocommerce-ios/pull/7886/commits/9139d171fa69b9dadb9a1626b2700f8f9084cac9